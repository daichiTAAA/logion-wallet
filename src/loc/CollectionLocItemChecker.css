.CollectionLocItemChecker {
    margin-top: 20px;
}

.CollectionLocItemChecker .Button {
    margin-left: 20px;
}

.CollectionLocItemChecker .IconTextRow .text {
    width: calc(100% - 50px);
}

.CollectionLocItemChecker .CheckResultFeedback {
    margin-top: 20px;
}

.CollectionLocItemChecker .CheckResultFeedback .label-positive {
    color: #37ad4b;
    font-weight: bold;
}

.CollectionLocItemChecker .CheckResultFeedback .label-negative {
    color: #ea1f46;
    font-weight: bold;
}

.CollectionLocItemChecker .CheckResultFeedback .Col {
    padding-right: 20px;
    min-width: 65px;
}

.CollectionLocItemChecker .CheckResultFeedback .url-header {
    font-weight: bold;
}

.CollectionLocItemChecker div.IconTextRow div.text div.FormGroup div div.Row div.Col {
    width: 50rem;
}

