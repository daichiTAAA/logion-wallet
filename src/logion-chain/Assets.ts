import { ApiPromise } from '@polkadot/api';
import { StorageKey, Option } from '@polkadot/types';
import {
    AssetId as PolkadotAssetId,
    AssetDetails as PolkadotAsset,
    AssetMetadata as PolkadotAssetMetadata,
    Call
} from '@polkadot/types/interfaces';
import { randomBytes } from 'crypto';
import BN from 'bn.js';
import {
    AssetId as AssetIdType,
    AssetBalance as AssetBalanceType,
    AssetMetadata as AssetMetadataType,
    Asset as AssetType,
    AssetWithBalance as AssetWithBalanceType,
} from './Types';

import {
    SignAndSendCallback,
    ErrorCallback,
    signAndSend,
    Unsubscriber
} from './Signature';

export const DEFAULT_ASSETS_DECIMALS = 18;

export interface AssetCreationParameters {
    signerId: string,
    callback: SignAndSendCallback,
    errorCallback: ErrorCallback,
    api: ApiPromise,
}

export type AssetId = AssetIdType;

export interface AssetCreationResult {
    assetId: AssetId,
    unsubscriber: Unsubscriber,
}

export async function createAsset(parameters: AssetCreationParameters): Promise<AssetCreationResult> {
    const {
        api,
        signerId,
        callback,
        errorCallback,
    } = parameters;

    const assetId = await nextAssetId(api);
    const assetCreate: any = api.tx.assets.create; // circumvent typing in order to add missing argument
    const unsubscriber = signAndSend({
        signerId,
        submittable: assetCreate(assetId, signerId, 10, 1),
        callback,
        errorCallback,
    });

    return {
        assetId,
        unsubscriber
    };
}

async function nextAssetId(api: ApiPromise): Promise<AssetId> {
    const BYTES_IN_ASSET_ID = 64 / 8;
    const value = randomBytes(BYTES_IN_ASSET_ID);
    let candidate = new BN(value.toString('hex'), 16);
    let candidateUsed = await assetExists(api, candidate);

    let iteration = 0;
    const MAX_ITERATIONS = 10;
    while(candidateUsed && iteration < MAX_ITERATIONS) {
        candidate = new BN(value.toString('hex'), 16);
        candidateUsed = await assetExists(api, candidate);
        ++iteration;
    }

    if(candidateUsed) {
        throw new Error("Unable to select an unused ID");
    } else {
        return candidate;
    }
}

async function assetExists(api: ApiPromise, id: AssetId): Promise<boolean> {
    const asset = await api.query.assets.asset(id);
    return asset.isSome;
}

export type AssetMetadata = AssetMetadataType;

export interface SetAssetMetadataParameters {
    signerId: string,
    callback: SignAndSendCallback,
    errorCallback: ErrorCallback,
    api: ApiPromise,
    assetId: AssetId,
    metadata: AssetMetadata,
}

export function setAssetMetadata(parameters: SetAssetMetadataParameters): Unsubscriber {
    const {
        api,
        signerId,
        callback,
        errorCallback,
        assetId,
        metadata,
    } = parameters;
    return signAndSend({
        signerId,
        submittable: api.tx.assets.setMetadata(assetId, metadata.name, metadata.symbol, metadata.decimals),
        callback,
        errorCallback,
    });
}
export type AssetBalance = AssetBalanceType;

export function balanceFromAmount(tokens: number, decimals: number): AssetBalance {
    let tokensBN = new BN(tokens);
    let decimalsBN = new BN(decimals);
    let ten = new BN(10);
    return tokensBN.mul(ten.pow(decimalsBN));
}

export interface MintParameters {
    signerId: string,
    beneficiary: string,
    callback: SignAndSendCallback,
    errorCallback: ErrorCallback,
    api: ApiPromise,
    assetId: AssetId,
    amount: AssetBalance
}

export function mintTokens(parameters: MintParameters): Unsubscriber {
    const {
        api,
        signerId,
        beneficiary,
        callback,
        errorCallback,
        assetId,
        amount,
    } = parameters;
    return signAndSend({
        signerId,
        submittable: api.tx.assets.mint(assetId, beneficiary, amount),
        callback,
        errorCallback,
    });
}

export function tokensFromBalance(balance: AssetBalance, decimals: number): string {
    let balanceBN = new BN(balance);
    let exponent = new BN(decimals);
    let ten = new BN(10);
    return balanceBN.div(ten.pow(exponent)).toString();
}

export interface AssetBalanceParameters {
    api: ApiPromise,
    assetId: AssetId,
    decimals: number,
    address: string,
}

export async function assetBalance(parameters: AssetBalanceParameters): Promise<string> {
    const {
        api,
        assetId,
        decimals,
        address,
    } = parameters;
    const balance = await api.query.assets.account(assetId, address);
    return tokensFromBalance(balance.balance, decimals);
}

export interface GetAssetsParameters {
    api: ApiPromise,
}

export type Asset = AssetType;

export async function getAssets(parameters: GetAssetsParameters): Promise<Asset[]> {
    const {
        api,
    } = parameters;
    const assetIds = extractAssetIds(await api.query.assets.asset.keys());
    const assetDetails: Option<PolkadotAsset>[] = await Promise.all(assetIds.map(assetId => api.query.assets.asset(assetId)));
    const metadata: PolkadotAssetMetadata[] = await Promise.all(assetIds.map(assetId => api.query.assets.metadata(assetId)));
    return assetIds.map((assetId, index) => {
        return {
            assetId,
            issuer: assetDetails[0].unwrap().issuer.toString(),
            metadata: {
                name: metadata[index].name.toUtf8(),
                symbol: metadata[index].symbol.toUtf8(),
                decimals: metadata[index].decimals.toNumber(),
            }
        };
    });
}

function extractAssetIds(keys: StorageKey<[PolkadotAssetId]>[]): AssetId[] {
    return keys.map(({ args: [assetId] }) => assetId);
}

export interface AccountBalanceParameters {
    api: ApiPromise,
    accountId: string,
    assets?: Asset[],
}

export type AssetWithBalance = AssetWithBalanceType;

export async function accountBalance(parameters: AccountBalanceParameters): Promise<AssetWithBalance[]> {
    const {
        api,
        accountId,
        assets,
    } = parameters;
    

    let theAssets;
    if(assets !== undefined) {
        theAssets = assets;
    } else {
        theAssets = await getAssets({api: api!});
    }
    const balances: string[] = await Promise.all(theAssets.map(asset => assetBalance({
        api,
        assetId: asset.assetId,
        decimals: asset.metadata.decimals,
        address: accountId
    })));
    return theAssets.map((asset, index) => ({
        asset,
        balance: balances[index],
    }));
}

export interface BuildTransferCallParameters {
    api: ApiPromise,
    assetId: AssetId,
    target: string,
    amount: AssetBalance,
}

export function buildTransferCall(parameters: BuildTransferCallParameters): Call {
    const {
        api,
        assetId,
        target,
        amount,
    } = parameters;

    return api.createType('Call', api.tx.assets.transfer(assetId, target, amount));
}
