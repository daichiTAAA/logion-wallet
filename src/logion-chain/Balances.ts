import { ApiPromise } from '@polkadot/api';
import { PrefixedNumber, ScientificNumber, convertToPrefixed, NONE, ATTO } from "./numbers";
import { ExtrinsicSubmissionParameters, Unsubscriber, signAndSend } from './Signature';
import { Call } from "@polkadot/types/interfaces";
import { SubmittableExtrinsic } from "@polkadot/api/submittable/types";

const LGNT_DECIMALS = 18;
export const LGNT_SMALLEST_UNIT = ATTO;
export const SYMBOL = "LGNT";

export interface GetAccountDataParameters {
    api: ApiPromise,
    accountId: string,
}

export interface AccountData {
    available: string,
    reserved: string,
    total: string,
}

export async function getAccountData(parameters: GetAccountDataParameters): Promise<AccountData> {
    const {
        api,
        accountId,
    } = parameters;

    const accountData = await api.query.system.account(accountId);

    return {
        available: accountData.data.free.toString(),
        reserved: accountData.data.reserved.toString(),
        total: accountData.data.free.add(accountData.data.reserved).toString(),
    };
}
export interface Coin {
    id: string,
    name: string,
    iconId: string,
    iconType: 'svg' | 'png',
    symbol: string,
}

export interface CoinBalance {
    coin: Coin,
    balance: PrefixedNumber,
    level: number,
}

export async function getBalances(parameters: GetAccountDataParameters): Promise<CoinBalance[]> {
    const {
        api,
        accountId,
    } = parameters;

    const data = await getAccountData({
        api: api!,
        accountId,
    });

    const logAvailable = scientificLogBalance(data.total).optimizeScale(3);
    const logPrefixedAvailable = convertToPrefixed(logAvailable);
    const logLevel = logAvailable.divideBy(ARTIFICIAL_MAX_BALANCE).toNumber();

    return [
        buildCoinBalance('lgnt', logPrefixedAvailable, logLevel),
        buildCoinBalance('dot', new PrefixedNumber("0", NONE), 1)
    ];
}

export function scientificLogBalance(tokens: string): ScientificNumber {
    return new ScientificNumber(tokens, -LGNT_DECIMALS).optimizeScale(3);
}

export function prefixedLogBalance(tokens: string): PrefixedNumber {
    const scientific = scientificLogBalance(tokens);
    return convertToPrefixed(scientific);
}

const ARTIFICIAL_MAX_BALANCE = scientificLogBalance("100");

function buildCoinBalance(coinId: string, balance: PrefixedNumber, level: number): CoinBalance {
    const coin = getCoin(coinId);
    return {
        coin,
        balance,
        level,
    }
}

export function getCoin(coinId: string): Coin {
    if(coinId === 'dot') {
        return {
            id: 'dot',
            name: 'Polkadot',
            iconId: 'dot',
            iconType: 'png',
            symbol: 'DOT',
        };
    } else if(coinId === "lgnt") {
        return {
            id: 'lgnt',
            name: 'logion Token',
            iconId: 'lgnt',
            iconType: 'svg',
            symbol: SYMBOL,
        };
    } else {
        throw new Error(`Unsupported coin ${coinId}`);
    }
}

export interface TransferParameters extends ExtrinsicSubmissionParameters, BuildTransferCallParameters {
}

export function transfer(parameters: TransferParameters): Unsubscriber {
    const {
        signerId,
        callback,
        errorCallback,
    } = parameters;

    return signAndSend({
        signerId,
        submittable: transferSubmittable(parameters),
        callback,
        errorCallback,
    });
}

export interface BuildTransferCallParameters {
    api: ApiPromise;
    destination: string;
    amount: PrefixedNumber;
}

function transferSubmittable(parameters: BuildTransferCallParameters): SubmittableExtrinsic<'promise'> {
    const {
        api,
        destination,
        amount
    } = parameters;
    return api.tx.balances.transfer(destination, amount.convertTo(LGNT_SMALLEST_UNIT).coefficient.unnormalize())
}

export function buildTransferCall(parameters: BuildTransferCallParameters): Call {
    return parameters.api.createType('Call', transferSubmittable(parameters))
}

export async function estimateFee(parameters: BuildTransferCallParameters): Promise<PrefixedNumber> {
    const submittable = transferSubmittable(parameters);
    const paymentInfo = await submittable.paymentInfo(parameters.destination);
    return new PrefixedNumber(paymentInfo.partialFee.toString(), LGNT_SMALLEST_UNIT);
}
