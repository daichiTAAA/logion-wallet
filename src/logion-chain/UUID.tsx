import { v4, parse, stringify } from 'uuid';

export class UUID {

    constructor(value?: string | Array<number>) {
        if(value === undefined) {
            this.bytes = parse(v4())
        } else if(typeof value === 'string') {
            this.bytes = parse(value);
        } else {
            this.bytes = [ ...value ];
        }
    }

    private bytes: ArrayLike<number>;

    toString(): string {
        return stringify(this.bytes);
    }

    toHexString(): string {
        return "0x" + this.toString().replace(/-/g, "");
    }
}
