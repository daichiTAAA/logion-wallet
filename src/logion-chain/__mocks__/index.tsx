import { Extension } from '../Keys';

const LogionChainContextProvider = (props: any) => null;

let context = {
    apiState: 'CONNECT_INIT',
    injectedAccounts: []
};

export function setContextMock(value: any) {
    context = value;
}

export function useLogionChain() {
    return context;
}

let extensionAvailable = false;

export function setExtensionAvailable(value: boolean) {
    extensionAvailable = value;
}

export function isExtensionAvailable() {
    return extensionAvailable;
}

let recommendedExtensionValue: Extension | null = null;

export function setRecommendedExtension(value: Extension | null) {
    recommendedExtensionValue = value;
}

export function recommendedExtension(): Extension | null {
    return recommendedExtensionValue;
}

export {
    LogionChainContextProvider,
};
