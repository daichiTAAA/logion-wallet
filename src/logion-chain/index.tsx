// Imports
import { KeyringPair as PolkadotKeyringPair } from '@polkadot/keyring/types';
import {
    Block as PolkadotBlock,
    Hash as PolkadotHash,
    Header as PolkadotHeader,
    SignedBlock as PolkadotSignedBlock,
    EventRecord as PolkadotEventRecord,
    Extrinsic as PolkadotExtrinsic,
} from '@polkadot/types/interfaces';
import keyring from '@polkadot/ui-keyring';
import { InjectedAccountWithMeta as PolkadotInjectedAccountWithMeta } from '@polkadot/extension-inject/types';
import { ISubmittableResult as PolkadotISubmittableResult } from '@polkadot/types/types';

import {
    LogionChainContextProvider,
    useLogionChain,

    ExtrinsicFetchSpecification as ContextExtrinsicFetchSpecification,
    ApiState as ContextApiState,
    NodeMetadata as ContextNodeMetadata,
} from './LogionChainContext';

import {
    signAndSend,
    replaceUnsubscriber,
    sign,

    ExtrinsicSignatureParameters as SignatureExtrinsicSignatureParameters,
    Unsubscriber as SignatureUnsubscriber,
    SignAndSendCallback as SignatureSignAndSendCallback,
    AttributesSignatureParameters as SignatureStringSignatureParameters,
} from './Signature';

import {
    getTimestamp,
    fetchExtrinsics,
    hashEquals,

    ExtrinsicsAndHead as BlocksExtrinsicsAndHead,
} from './Blocks';

import {
    isExtensionAvailable,
    recommendedExtension
} from './Keys';

// Re-exports
export {
    useLogionChain,
    LogionChainContextProvider,
    getTimestamp,
    fetchExtrinsics,
    hashEquals,
    signAndSend,
    replaceUnsubscriber,
    isExtensionAvailable,
    recommendedExtension,
    keyring,
    sign,
};

export type Unsubscriber = SignatureUnsubscriber;
export type ExtrinsicSignatureParameters = SignatureExtrinsicSignatureParameters;
export type ExtrinsicFetchSpecification = ContextExtrinsicFetchSpecification;

export type ExtrinsicsAndHead = BlocksExtrinsicsAndHead;

export type KeyringPair = PolkadotKeyringPair;
export type Hash = PolkadotHash;
export type SignedBlock = PolkadotSignedBlock;
export type Header = PolkadotHeader;
export type Block = PolkadotBlock;
export type EventRecord = PolkadotEventRecord;
export type Extrinsic = PolkadotExtrinsic;
export type SignAndSendCallback = SignatureSignAndSendCallback;
export type ApiState = ContextApiState;
export type NodeMetadata = ContextNodeMetadata;
export type InjectedAccountWithMeta = PolkadotInjectedAccountWithMeta;
export type StringSignatureParameters = SignatureStringSignatureParameters;
export type ISubmittableResult = PolkadotISubmittableResult;
