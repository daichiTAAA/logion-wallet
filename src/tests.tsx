import { ReactElement } from 'react';
import { createRenderer } from 'react-test-renderer/shallow';
import reactRenderer, { ReactTestRenderer, ReactTestRendererJSON, act as reactAct } from 'react-test-renderer';

export function render(element: ReactElement): null | ReactTestRendererJSON | ReactTestRendererJSON[] {
    const component = reactRenderer.create(element);
    return component.toJSON();
}

export function renderer(element: ReactElement): ReactTestRenderer {
    return reactRenderer.create(element);
}

export function shallowRender<E extends ReactElement>(element: ReactElement): E {
    const renderer = createRenderer();
    renderer.render(element);
    return renderer.getRenderOutput();
}

export function mockAccount(address: string, name: string) {
    return {
        address,
        meta: {
            name
        }
    };
}

export { reactAct as act };
