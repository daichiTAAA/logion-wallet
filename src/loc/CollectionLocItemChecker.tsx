import { UUID } from "../logion-chain/UUID";
import React, { useState, useCallback, useEffect } from "react";
import { useLogionChain } from "../logion-chain";
import { getCollectionItem, getCollectionSize } from "../logion-chain/LogionLoc";
import PolkadotFrame from "../common/PolkadotFrame";
import { useCommonContext } from "../common/CommonContext";
import Icon from "../common/Icon";
import IconTextRow from "../common/IconTextRow";
import FormGroup from "../common/FormGroup";
import { Form } from "react-bootstrap";
import Button from "../common/Button";
import { Row, Col } from "../common/Grid";
import CopyPasteButton from "../common/CopyPasteButton";
import { fullCollectionItemCertificate } from "../PublicPaths";
import "./CollectionLocItemChecker.css"

export interface Props {
    locId: UUID,
}

export type CheckResult = 'NONE' | 'POSITIVE' | 'NEGATIVE';

export default function CollectionLocItemChecker(props: Props) {

    const { colorTheme } = useCommonContext();
    const { locId } = props;
    const [ state, setState ] = useState<CheckResult>('NONE')
    const [ collectionSize, setCollectionSize ] = useState<number | undefined | null>(null);
    const [ itemId, setItemId ] = useState<string>("")
    const { api } = useLogionChain();

    useEffect(() => {
        if (api && collectionSize === null) {
            getCollectionSize({ api, locId })
                .then(setCollectionSize)
        }
    }, [ api, locId, collectionSize ])

    const checkData = useCallback(async () => {
        if (api) {
            try {
                const collectionItem = await getCollectionItem({ api, locId, itemId })
                if (collectionItem) {
                    setState('POSITIVE')
                } else {
                    setState('NEGATIVE')
                }
            } catch (e) {
                setState('NEGATIVE')
            }
        }
    }, [ api, locId, itemId ]);

    return (
        <PolkadotFrame className="CollectionLocItemChecker" colorTheme={ colorTheme }>
            <IconTextRow
                icon={ <Icon icon={ { id: "polkadot_collection" } } width="45px" /> }
                text={ <>
                    <p className="text-title">Number of Collection-related data recorded on logion
                        blockchain: { collectionSize }</p>
                    <p>The Collection LOC material listed above benefits all data imported through the logion API by an
                        external application approved between the Legal Officer and its client under a process validated
                        by the Legal Officer of the present LOC.</p>
                    <p>To check if a data is covered by this Collection LOC and get its online public certificate, just
                        submit the related data in the input field below:</p>
                    <FormGroup
                        id="itemId"
                        noFeedback={ true }
                        control={
                            <Row>
                                <Col>
                                    <Form.Control
                                        type="text"
                                        value={ itemId }
                                        onChange={ value => {
                                            setState('NONE')
                                            setItemId(value.target.value);
                                        }}
                                    />
                                </Col>
                                <Col>
                                    <Button onClick={ checkData }>Check data</Button>
                                </Col>
                            </Row>
                        }
                        colors={ colorTheme.dialog }
                    />
                    <CheckResultFeedback locId={ locId } itemId={ itemId } state={ state } />
                </>
                } />
        </PolkadotFrame>)
}

interface CheckResultProps {
    locId: UUID,
    itemId: string,
    state: CheckResult,
}

function CheckResultFeedback(props: CheckResultProps) {
    const { locId, itemId, state } = props
    switch (state) {
        case "POSITIVE":
            const certificateUrl = fullCollectionItemCertificate(locId, itemId)
            return (
                <Row className="CheckResultFeedback result-positive" dataTestId={ `feedback-${ state }` }>
                    <Col>
                        <p>
                            Check result: <span className="label-positive">positive</span><br />
                            The data you uploaded is covered by the current Collection LOC.
                        </p>
                    </Col>
                    <Col>
                        <Icon icon={ { id: "ok" } } height='45px' />
                    </Col>
                    <Col>
                        <p>
                            <span className="url-header">Certificate Public web address (URL) for the data covered by this Collection LOC:</span><br />
                            <a href={ certificateUrl } target="_blank" rel="noreferrer">{ certificateUrl }</a>
                            <CopyPasteButton value={ certificateUrl } />
                        </p>
                    </Col>
                </Row>
            )
        case "NEGATIVE":
            return (
                <Row className="CheckResultFeedback result-negative" dataTestId={ `feedback-${ state }` }>
                    <Col>
                        <p>
                            Check result: <span className="label-negative">negative</span><br />
                            The data you uploaded has no match and is NOT covered by the current<br />
                            Collection LOC. Please be careful and execute a deeper due diligence.
                        </p>
                    </Col>
                    <Col>
                        <Icon icon={ { id: "ko" } } height='45px' />
                    </Col>
                </Row>
            )
        case "NONE":
            return (
                <Row className="CheckResultFeedback result-none" children="" dataTestId={ `feedback-${ state }` } />
            )
    }
}
