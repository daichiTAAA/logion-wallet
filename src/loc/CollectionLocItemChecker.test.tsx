jest.mock("../common/CommonContext");
jest.mock("../logion-chain");

import { findByTestId, getByTestId } from '@testing-library/react';
import CollectionLocItemChecker from "./CollectionLocItemChecker";
import { UUID } from "../logion-chain/UUID";
import { clickByName, render } from "../tests";
import { act } from "react-dom/test-utils";
import ReactDOM from "react-dom";

fdescribe("CollectionLocItemChecker", () => {

    it("renders", () => {
        const locId: UUID = new UUID("d97c99fd-9bcc-4f92-b9ea-b6be93abbbcd")
        const result = render(<CollectionLocItemChecker locId={ locId } />);
        expect(result).toMatchSnapshot();
    })

    it("clicks on Check data", async () => {

        const container = document.createElement('div');
        document.body.appendChild(container);

        const locId: UUID = new UUID("d97c99fd-9bcc-4f92-b9ea-b6be93abbbcd")
        ReactDOM.render(<CollectionLocItemChecker locId={ locId } />, container);

        await act(async () => {
            await clickByName("Check data");
        })
        expect(await getByTestId(container, "feedback-POSITIVE")).toBeDefined()
        expect(() => getByTestId(container, "feedback-NEGATIVE")).toThrowError()
        expect(() => getByTestId(container, "feedback-NONE")).toThrowError()
    })
})
