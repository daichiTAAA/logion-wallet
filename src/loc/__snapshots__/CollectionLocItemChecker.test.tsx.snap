// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CollectionLocItemChecker renders 1`] = `
<div
  className="Frame PolkadotFrame CollectionLocItemChecker"
  style={
    Object {
      "backgroundColor": "#ffffff",
      "boxShadow": "0 0 25px #3b6cf41a",
      "color": "#000000",
    }
  }
>
  <style>
    
    .Frame a,
    .Frame .btn-link {
        color: #3b6cf426
    }
    
  </style>
  <div
    className="IconTextRow"
  >
    <div
      className="icon"
    >
      <img
        alt=""
        src="/assets/polkadot_collection.svg"
        width="45px"
      />
    </div>
    <div
      className="text"
    >
      <p
        className="text-title"
      >
        Number of Collection-related data recorded on logion blockchain: 
      </p>
      <p>
        The Collection LOC material listed above benefits all data imported through the logion API by an external application approved between the Legal Officer and its client under a process validated by the Legal Officer of the present LOC.
      </p>
      <p>
        To check if a data is covered by this Collection LOC and get its online public certificate, just submit the related data in the input field below:
      </p>
      <div
        className="FormGroup"
        style={
          Object {
            "backgroundColor": "#eff3fe",
          }
        }
      >
        <style>
          
    .FormGroup .form-control,
    .FormGroup .form-control[readonly],
    .FormGroup .input-group .btn-primary {
        color: #000000;
    }
    
        </style>
        <div
          data-testid="itemIdGroup"
        >
          <div
            className="Row"
          >
            <div
              className="Col"
            >
              <input
                className="form-control"
                id="itemId"
                onChange={[Function]}
                type="text"
                value=""
              />
            </div>
            <div
              className="Col"
            >
              <button
                className="Button btn btn-primary"
                disabled={false}
                onClick={[Function]}
                style={Object {}}
                type="button"
              >
                Check data
              </button>
            </div>
          </div>
        </div>
      </div>
      <div
        className="Row CheckResultFeedback result-none"
        data-testid="feedback-NONE"
      >
        
      </div>
    </div>
  </div>
</div>
`;
